janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java

```diff
@@ -241,7 +262,29 @@ public static boolean isEmpty(Condition<?> condition) {
     private static <E extends JanusGraphElement> Or<E> addConstraint(final RelationType type, OrJanusPredicate predicate, List<Object> values, Or<E> or, StandardJanusGraphTx tx) {
         for (int i = 0 ; i < values.size(); i++) {
             final JanusGraphPredicate janusGraphPredicate = predicate.get(i);
-            if (janusGraphPredicate instanceof AndJanusPredicate) {
+            if (janusGraphPredicate instanceof Contain) {
+                //Rewrite contains conditions
+                final Collection childValues = (Collection) values.get(i);
+                if (janusGraphPredicate == Contain.NOT_IN) {
+                    if (childValues.isEmpty()) continue; //Simply ignore since trivially satisfied
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
+                } else {
+                    Preconditions.checkArgument(janusGraphPredicate == Contain.IN);
+                    if (childValues.isEmpty()) {
+                        return null; //Cannot be satisfied
```

Comment:
This would fail the test
```java
assertEquals(3, g.V().or(__.has("id", P.within(2, 3, 4)), __.has("id", P.within())).count().next());
```

Reply:
Thanks for the review. Basically needed to change two lines within the definition of `OR` to catch that bug. I updated the branch by pushing a separate commit containing the fix so it is easier to review.

Also added 14 new test cases which cover all of the empty argument cases as well as some previously untested combinations of `with` and `without` cases. If you can find different combinations worth testing, please let me know.